File: AIReaderKindle/src/Domain/Cards/MatchRound.cpp
#include "MatchRound.hpp"

#include <algorithm>
#include <numeric>
#include <random>

namespace {

bool anyRowAligned(const std::vector<int>& fronts, const std::vector<int>& backs) {
    for (size_t i = 0; i < fronts.size(); ++i) if (fronts[i] == backs[i]) return true;
    return false;
}

}  // namespace

MatchRound::MatchRound(std::vector<Card> cards, unsigned seed)
    : cards_(std::move(cards)), matched_(cards_.size(), false) {
    fronts_.resize(cards_.size());
    std::iota(fronts_.begin(), fronts_.end(), 0);
    backs_ = fronts_;
    std::mt19937 random(seed);
    std::shuffle(fronts_.begin(), fronts_.end(), random);
    // A row that lines up would give its answer away.
    do {
        std::shuffle(backs_.begin(), backs_.end(), random);
    } while (cards_.size() > 1 && anyRowAligned(fronts_, backs_));
}

int MatchRound::matchedCount() const {
    return static_cast<int>(std::count(matched_.begin(), matched_.end(), true));
}

bool MatchRound::canPick(int card) const {
    return card >= 0 && card < static_cast<int>(cards_.size()) && !matched_[card];
}

std::optional<MatchRound::Pick> MatchRound::pickFront(int card) {
    lastPick_.reset();
    if (!canPick(card)) return std::nullopt;
    if (front_ == card) front_.reset(); else front_ = card;
    return resolve();
}

std::optional<MatchRound::Pick> MatchRound::pickBack(int card) {
    lastPick_.reset();
    if (!canPick(card)) return std::nullopt;
    if (back_ == card) back_.reset(); else back_ = card;
    return resolve();
}

std::optional<MatchRound::Pick> MatchRound::resolve() {
    if (!front_ || !back_) return std::nullopt;
    Pick pick{*front_, *back_, *front_ == *back_};
    if (pick.matched) matched_[pick.front] = true; else ++misses_;
    front_.reset();
    back_.reset();
    lastPick_ = pick;
    return pick;
}

File: AIReader/AIReader/Domain/Cards/MatchRound.swift
import Foundation

/// One round of the matching game: a few cards, their fronts in one column
/// and their backs in another, shuffled apart. The reader taps a front and a
/// back; a pair that belongs together is put aside, one that does not is a
/// miss. Pure state, so it can be checked without a screen.
struct MatchRound: Equatable, Sendable {
    /// A front and a back the reader put together, as indices into `cards`.
    struct Pick: Equatable, Sendable {
        let front: Int
        let back: Int
        let matched: Bool
    }

    let cards: [Card]
    /// The order the fronts and backs are shown in, as indices into `cards`.
    /// No back sits in the same row as its front.
    private(set) var fronts: [Int]
    private(set) var backs: [Int]
    private var matched: [Bool]
    private(set) var selectedFront: Int?
    private(set) var selectedBack: Int?
    /// The pair most recently put together, until the next tap.
    private(set) var lastPick: Pick?
    private(set) var misses = 0

    init(cards: [Card], using generator: inout some RandomNumberGenerator) {
        self.cards = cards
        matched = Array(repeating: false, count: cards.count)
        fronts = Array(cards.indices).shuffled(using: &generator)
        backs = fronts
        // A row that lines up would give its answer away.
        repeat {
            backs.shuffle(using: &generator)
        } while cards.count > 1 && zip(fronts, backs).contains { $0 == $1 }
    }

    func isMatched(_ card: Int) -> Bool { matched[card] }
    var matchedCount: Int { matched.filter { $0 }.count }
    var isComplete: Bool { matchedCount == cards.count }

    /// A tap on a front or a back. Tapping the chosen one again lets go of
    /// it; once one of each is chosen the pair resolves and is returned.
    @discardableResult
    mutating func pickFront(_ card: Int) -> Pick? {
        lastPick = nil
        guard canPick(card) else { return nil }
        selectedFront = selectedFront == card ? nil : card
        return resolve()
    }

    @discardableResult
    mutating func pickBack(_ card: Int) -> Pick? {
        lastPick = nil
        guard canPick(card) else { return nil }
        selectedBack = selectedBack == card ? nil : card
        return resolve()
    }

    private func canPick(_ card: Int) -> Bool {
        cards.indices.contains(card) && !matched[card]
    }

    private mutating func resolve() -> Pick? {
        guard let front = selectedFront, let back = selectedBack else { return nil }
        let pick = Pick(front: front, back: back, matched: front == back)
        if pick.matched { matched[front] = true } else { misses += 1 }
        selectedFront = nil
        selectedBack = nil
        lastPick = pick
        return pick
    }
}
